### PR TITLE
feat(tests): introduces tests against MicroK8s

### LINKED ISSUE #322: Test on vanilla Kubernetes/Istio

### PR DESCRIPTION
**Pull request by bartoszmajsak**
#### Short description of what this resolves:

Ensures that istio-workspace works against vanilla k8s with istio by running tests on [MicroK8s](https://microk8s.io)

#### Changes proposed in this pull request:

- new Circle CI job running tests against MicroK8s
- required changes in the test code
- docs: dev guide update

Fixes #322

### CODE REVIEW

File: deploy/operator.yaml [hunk 1]
```diff
@@ -35,7 +35,7 @@ spec:
                   fieldPath: metadata.name
             - name: OPERATOR_NAME
               value: istio-workspace
-          image: quay.io/maistra/istio-workspace:v0.0.3
+          image: localhost:32000/istio-workspace-images/istio-workspace:latest
```
Comment (aslakknutsen) [inline deploy/operator.yaml:38]: you sure? :)

File: e2e/operator_installation_test.go [hunk 2]
```diff
@@ -55,6 +55,13 @@ var _ = Describe("Operator Installation Tests", func() {
 		})
 
 		It("should install into current namespace", func() {
+			if !RunsAgainstOpenshift {
+				Skip("OpenShift specific test assuming current namespace is set and oc available")
```
Comment (aslakknutsen) [inline e2e/operator_installation_test.go:59]: Add some notes around "skip" explaining how this is covered else where or similar if it's no a pure Openshift test

File: e2e/smoke_test.go [hunk 3]
```diff
@@ -76,39 +81,7 @@ var _ = Describe("Smoke End To End Tests - against OpenShift Cluster with Istio
 							"--run", "ruby ratings.rb 9080",
 							"--route", "header:x-test-suite=smoke",
 							"--session", sessionName,
-						)
```
Comment (aslakknutsen) [inline e2e/smoke_test.go:79]: What happened here?

File: pkg/internal/session/session.go [hunk 4]
```diff
@@ -177,7 +177,7 @@ func (h *handler) waitForRefToComplete() (*istiov1alpha1.Session, string, error)
 	if h.opts.Duration != nil {
 		duration = *h.opts.Duration
 	}
-	err = wait.Poll(duration, 10*time.Second, func() (bool, error) {
+	err = wait.Poll(duration, 60*time.Second, func() (bool, error) {
```
Comment (aslakknutsen) [inline pkg/internal/session/session.go:180]: microk8s slower?

File: deploy/operator.yaml (hunk 1 above)
Comment (bartoszmajsak) [reply deploy/operator.yaml]: That is autogenerated and already reverted

File: pkg/internal/session/session.go (hunk 4 above)
Comment (bartoszmajsak) [reply pkg/internal/session/session.go:180]: Yeah, it looked like that yesterday. Was wondering the same. Maybe I will bring it back to 10.

```suggestion
	err = wait.Poll(duration, 10*time.Second, func() (bool, error) {
```

File: e2e/operator_installation_test.go (hunk 2 above)
Comment (bartoszmajsak) [reply e2e/operator_installation_test.go:59]: Where do you think we should add that? In the message itself, docs or ...?

File: e2e/smoke_test.go (hunk 3 above)
Comment (bartoszmajsak) [reply e2e/smoke_test.go:79]: Removed the test which is redundant. This is covered by `DeploymentConfig` anyway (implicit namespace).

File: .circleci/config.yml [hunk 5]
```diff
@@ -142,6 +143,129 @@ jobs:
           paths:
             - ./e2e-openshift-4.githash
 
+  ## End-to-end testing using single-node microk8s cluster
+  e2e_tests_microk8s:
+    working_directory: ~/.go_workspace/src/github.com/maistra/istio-workspace
+#    machine: true
```
Comment (bartoszmajsak) [inline .circleci/config.yml:149]:
```suggestion
```

File: deploy/olm-catalog/istio-workspace/manifests/istio-workspace.clusterserviceversion.yaml [hunk 6]
```diff
@@ -25,7 +25,7 @@ metadata:
         }
       ]
     capabilities: Basic Install
-  name: istio-workspace.v0.0.3
+  name: istio-workspace.v0.0.3-next-554fd4f-dirty-next-554fd4f-dirty
```
Comment (bartoszmajsak) [inline deploy/olm-catalog/istio-workspace/manifests/istio-workspace.clusterserviceversion.yaml]:
```suggestion
  name: istio-workspace.v0.0.3
```

File: deploy/olm-catalog/istio-workspace/manifests/istio-workspace.clusterserviceversion.yaml [hunk 7]
```diff
@@ -119,13 +119,13 @@ spec:
           selector:
             matchLabels:
               app: istio-workspace
-              version: v0.0.3
+              version: v0.0.3-next-554fd4f-dirty-next-554fd4f-dirty
```
Comment (bartoszmajsak) [inline deploy/olm-catalog/istio-workspace/manifests/istio-workspace.clusterserviceversion.yaml]:
```suggestion
              version: v0.0.3
```

File: deploy/olm-catalog/istio-workspace/manifests/istio-workspace.clusterserviceversion.yaml [hunk 8]
```diff
@@ -249,4 +249,5 @@ spec:
   maturity: alpha
   provider:
     name: Red Hat, Inc.
+  replaces: istio-workspace.v0.0.3-next-554fd4f-dirty-next-554fd4f-dirty
```
Comment (bartoszmajsak) [inline deploy/olm-catalog/istio-workspace/manifests/istio-workspace.clusterserviceversion.yaml:252]:
```suggestion
```

File: e2e/operator_installation_test.go (hunk 2 above)
Comment (aslakknutsen) [reply e2e/operator_installation_test.go:59]: A comment to the Skip line would do

File: e2e/smoke_test.go (hunk 3 above)
Comment (aslakknutsen) [reply e2e/smoke_test.go:79]: Now we have Deployment testing without NS and DeploymentConfig testing with NS? That would technically leave k8s without any tests for specific ns...?

Not the biggest chance of error in the world I guess but

File: pkg/assets/isto-workspace-deploy.go [hunk 9]
```diff
@@ -129,7 +129,7 @@ func deployService_accountYaml() (*asset, error) {
 	return a, nil
```
Comment (aslakknutsen) [inline pkg/assets/isto-workspace-deploy.go:129]: This should be left out of the commit as well, no?

File: e2e/smoke_test.go (hunk 3 above)
Comment (bartoszmajsak) [reply e2e/smoke_test.go:79]: Quite the opposite. We test `Deployment` with specified ns as we use kubectl for that, and current namespace/project (or implicit ns) for `DeploymentConfig`. Or am I missing sth?

File: pkg/internal/session/session.go (hunk 4 above)
Comment (bartoszmajsak) [inline pkg/internal/session/session.go:180]:
```suggestion
	err = wait.Poll(duration, 16*time.Second, func() (bool, error) {
```

File: e2e/smoke_test.go (hunk 3 above)
Comment (aslakknutsen) [reply e2e/smoke_test.go:79]: opposite, right still 1 option not test on k8s. but meh..

File: e2e/operator_installation_test.go (hunk 2 above)
Comment (bartoszmajsak) [reply e2e/operator_installation_test.go]: 598cd32

File: pkg/assets/isto-workspace-deploy.go (hunk 9 above)
Comment (bartoszmajsak) [reply pkg/assets/isto-workspace-deploy.go]: 434177f

Comment (bartoszmajsak) [reply pkg/assets/isto-workspace-deploy.go]: 434177f

**Pull request merged**